sdk/webpubsub/azure-messaging-webpubsubclient/azure/messaging/webpubsubclient/aio/_client.py

```diff
@@ -0,0 +1,1070 @@
+    CallbackType,
+    WebPubSubProtocolType,
+)
+from .._client import (
+    _WebPubSubClient,
+    WebPubSubClientCredential,
+    _RETRY_TOTAL,
+    _RETRY_BACKOFF_FACTOR,
+    _RETRY_BACKOFF_MAX,
+    _RECOVERY_TIMEOUT,
+    _RECOVERY_RETRY_INTERVAL,
+    _USER_AGENT,
+    _ACK_TIMEOUT,
+    _START_TIMEOUT,
+)
+from .._util import format_user_agent, raise_for_empty_message_ack, NO_ACK_MESSAGE_ERROR
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class WebSocketAppAsync:  # pylint: disable=too-many-instance-attributes
+    def __init__(
+        self,
+        url,
+        on_open=None,
+        on_message=None,
+        on_close=None,
+        subprotocols: Optional[List[str]] = None,
+        header: Optional[Dict[str, str]] = None,
+    ) -> None:
+        self.url = url
+        self.on_open = on_open
+        self.on_message = on_message
+        self.on_close = on_close
+        self.subprotocols = subprotocols
+        self.header = header
+        self.session: Optional[aiohttp.ClientSession] = None
+        self.sock: Optional[aiohttp.client._WSRequestContextManager] = None
+        self.event: asyncio.Event = asyncio.Event()
+        self.close_timeout = 1.0
+        self.keep_running = True
+
+    async def send(self, message: str) -> None:
+        await self.sock.send_str(message)  # type: ignore
+
+    async def run_forever(self):
+        try:
+            self.session = aiohttp.ClientSession()
+            self.sock = await self.session.ws_connect(self.url, protocols=self.subprotocols, headers=self.header)
+        except aiohttp.ClientConnectorError as e:
+            await self.close()
+            raise OpenClientError("Fail to open client") from e
+        await self.on_open()
+
+        self.keep_running = True
+        while self.keep_running:
+            msg: aiohttp.WSMessage = await self.sock.receive()  # type: ignore
+            if msg.type == aiohttp.WSMsgType.TEXT:
+                await self.on_message(msg.data)
+            elif msg.type == aiohttp.WSMsgType.CLOSING:
+                _LOGGER.debug("WebSocket is closing")
+            elif msg.type in (aiohttp.WSMsgType.CLOSED, aiohttp.WSMsgType.CLOSE):
+                _LOGGER.debug("WebSocket is closed")
+                await self.on_close(msg.data, msg.extra)
+                self.event.set()
+                break
+            elif msg.type == aiohttp.WSMsgType.ERROR:
+                await self.on_close(1008, "unexpected WebSocket error")
+                self.event.set()
+                break
+            else:
+                _LOGGER.debug("Unknown message: %s", msg)
+
+    async def connect(self):
+        self.session = aiohttp.ClientSession()
+        self.sock = await self.session.ws_connect(self.url, protocols=self.subprotocols, headers=self.header)
+
+    async def close(self):
+        try:
+            await self.sock.close()
+            # wait for on_close triggered
+            await asyncio.wait_for(self.event.wait(), self.close_timeout)
```

Comment:
in the case of a regular close of webpubsubclient, wont this event trigger an exception and write a log `Fail to close websocket connection` because `on_close` wont be called ?



for example

```python

service_client = WebPubSubServiceClient.from_connection_string(  # type: ignore
        connection_string=os.getenv("WEBPUBSUB_CONNECTION_STRING", ""), hub="hub"
    )

# do some work, its successful



service_client.close()

```

Reply:
For regular close, on_close will always be triggered, we have [testcase](https://github.com/Azure/azure-sdk-for-python/blob/a309d8fdd6d8b95a792e8f33d5920fa1f42bfcd5/sdk/webpubsub/azure-messaging-webpubsubclient/tests/test_smoke_async.py#L59-L75) for this scenario. There you can also try https://github.com/Azure/azure-sdk-for-python/blob/webpubsub-async-dev1/sdk/webpubsub/azure-messaging-webpubsubclient/samples/hello_world_async.py